aiokafka/record/legacy_records.py

```diff
@@ -278,155 +285,154 @@ def __repr__(self):
 
     def append(self, offset, timestamp, key, value):
         """ Append message to batch.
         """
-        # Check types
-        if type(offset) != int:
-            raise TypeError(offset)
         if self._magic == 0:
             timestamp = -1
         elif timestamp is None:
             timestamp = int(time.time() * 1000)
-        elif type(timestamp) != int:
-            raise TypeError(timestamp)
-        if not (key is None or
-                isinstance(key, (bytes, bytearray, memoryview))):
-            raise TypeError(
-                "Not supported type for key: {}".format(type(key)))
-        if not (value is None or
-                isinstance(value, (bytes, bytearray, memoryview))):
-            raise TypeError(
-                "Not supported type for value: {}".format(type(value)))
-
-        # Check if we have room for another message
-        pos = len(self._buffer)
-        size = self.size_in_bytes(offset, timestamp, key, value)
-        # We always allow at least one record to be appended
-        if offset != 0 and pos + size >= self._batch_size:
-            return None
-
-        # Allocate proper buffer length
-        self._buffer.extend(bytearray(size))
 
-        # Encode message
-        crc = self._encode_msg(pos, offset, timestamp, key, value)
+        # calculating length is not cheap; only do it once
+        key_size = len(key) if key is not None else 0
+        value_size = len(value) if value is not None else 0
 
-        return LegacyRecordMetadata(offset, crc, size, timestamp)
-
-    def _encode_msg(self, start_pos, offset, timestamp, key, value,
-                    attributes=0):
+        pos = self._pos
+        size = self.size_in_bytes(offset, timestamp, key_size, value_size)
+        # always allow at least one record to be appended
+        if offset != 0 and pos + size >= self._batch_size:
+            return None
+        elif offset == 0 and size > self._batch_size:
+            # if adding the message fails for any reason, this buffer will
+            # remain in its expanded state. Since position is tracked
+            # independently and comparisons are made against self._batch_size,
+            # this will have no ill effects
+            self._buffer = bytearray(size)
+
+        try:
+            crc = self._encode_msg(
+                pos, size, offset, timestamp, key_size, key, value_size, value)
+            self._pos += size
+            return LegacyRecordMetadata(offset, crc, size, timestamp)
+
+        except struct.error:
+            # perform expensive type checking only to translate struct errors
+            # to human-readable messages
+            if type(offset) != int:
+                raise TypeError(offset)
+            if type(timestamp) != int:
+                raise TypeError(timestamp)
+            if not isinstance(key, (bytes, bytearray, memoryview, NoneType)):
+                raise TypeError("Unsupported type for key: %s" % type(key))
```

Comment:
Not so sure about if this is a good idea. We perform len() call on those. It will raise a different error.

Reply:
True, but that error is already very explicit:
```python
>>> builder = LegacyRecordBatchBuilder(1, None, 16384)
>>> builder.append(0, None, object(), b"value")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/shawnh/repos/aiokafka/aiokafka/record/legacy_records.py", line 304, in append
    key_size = len(key) if key is not None else 0
TypeError: object of type 'object' has no len()
>>> builder.append(0, None, b"key", object())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/shawnh/repos/aiokafka/aiokafka/record/legacy_records.py", line 305, in append
    value_size = len(value) if value is not None else 0
TypeError: object of type 'object' has no len()
```

Python developers should be used to this pattern since type-checking is so expensive in the language. After all, most of the stdlib doesn't bother to check types and just ignores bad values or bubbles up esoteric errors when they're sent around.